### PR TITLE
Add shortcuts feature

### LINKED ISSUE #9: As a GP Clinic Management Staff who is impatient I can add new keyboard shortcuts to the application
... so that I can save time when performing frequently repeated tasks.

### LINKED ISSUE #94: Incorrect Handling of Empty Prefix Argument for Find Command
# Description
**Input:**
`find -pa n/`
<br>
**Expected Output:** 
Should show useful error messages indicating to user that prefix arguments should not be empty. Alternatively, should allow for empty String input argument for prefixes.
<br>
**Command Result:** 
No Changes. Command not parsed and executed successfully
<br>
**Notes:**
Exception thrown from `StringUtil::wordsContainSubstringIgnoreCase`

### PR DESCRIPTION
**Pull request by alvinlim277**
- [x] Adding shortcuts
- [x] Removing shortcuts
- [x] Saving shortcuts between sessions
- [x] Reduce messy dependencies
- [x] Clean up code
- [x] Update UG
- [x] Update Tests

Resolves #9 , resolve #94

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddShortcutCommand.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    public AddShortcutCommand(ShortcutAlias shortcutAlias, CommandWord commandWord) {
+        requireAllNonNull(shortcutAlias, commandWord);
+        this.shortcutAlias = shortcutAlias;
+        this.command = commandWord;
+    }
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        requireNonNull(model);
+        String feedback = model.getShortcutSettings().registerShortcut(shortcutAlias, command);
+        if (feedback == null) {
```
Comment (Gavino3o) [inline src/main/java/seedu/address/logic/commands/AddShortcutCommand.java:42]: Great use of feedback LinkedHashMap return value as feedback for override scenario!

File: src/main/java/seedu/address/model/UserPrefs.java [hunk 2]
```diff
@@ -35,6 +38,9 @@ public UserPrefs(ReadOnlyUserPrefs userPrefs) {
     public void resetData(ReadOnlyUserPrefs newUserPrefs) {
         requireNonNull(newUserPrefs);
         setGuiSettings(newUserPrefs.getGuiSettings());
+        ShortcutSettings newShortcutSettings = newUserPrefs.getShortcutSettings()
+                .removeBadMappings();
+        setShortcutSettings(newShortcutSettings);
```
Comment (Gavino3o) [inline src/main/java/seedu/address/model/UserPrefs.java:43]: Good addition!

File: src/main/java/seedu/address/logic/parser/AddShortcutCommandParser.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+        return new AddShortcutCommand(shortcutAlias, commandWord);
+    }
+    /**
+     * Returns true if none of the prefixes contains empty {@code Optional} values in the given
+     * {@code ArgumentMultimap}.
+     */
+    private static boolean arePrefixesPresent(ArgumentMultimap argumentMultimap, Prefix... prefixes) {
+        return Stream.of(prefixes).allMatch(prefix -> argumentMultimap.getValue(prefix).isPresent());
+    }
```
Comment (Gavino3o) [inline src/main/java/seedu/address/logic/parser/AddShortcutCommandParser.java:48]: This method can be moved into `ArgumentMultimap` class.

**Review by Gavino3o — COMMENTED** [review]
Overall LGTM. Just some minor nitpicks.

File: src/main/java/seedu/address/logic/parser/AddShortcutCommandParser.java (hunk 3 above)
Comment (alvinlim277) [reply src/main/java/seedu/address/logic/parser/AddShortcutCommandParser.java]: Agreed, will remove the copy of this method from `AddCommandParser` as well

File: src/main/java/seedu/address/logic/parser/ArgumentMultimap.java [hunk 4]
```diff
@@ -84,4 +84,11 @@ public static boolean arePrefixesPresent(ArgumentMultimap argumentMultimap, Pref
         return Stream.of(prefixes).allMatch(prefix -> argumentMultimap.getValue(prefix).isPresent());
     }
 
+    /**
+     * Returns true if none of the prefixes contains empty {@code Optional} values in the given
```
Comment (Gavino3o) [inline src/main/java/seedu/address/logic/parser/ArgumentMultimap.java:88]: I think you meant `Returns true if any of the prefixes` instead of `none` here.

**Review by Gavino3o — APPROVED** [review]

**Pull request merged**